java/sage/Carny.java

```diff
@@ -953,188 +1154,445 @@ private void stdProcessing()
-      if (a.time < currTime + currLookahead &&
-          a.time >= currTime - Scheduler.SCHEDULING_LOOKBEHIND && a.isTV())
+      if (a.getStartTime() < currTime + currLookahead &&
+          a.getStartTime() >= currTime - Scheduler.SCHEDULING_LOOKBEHIND && a.isTV())
+      {
         airset.add(a);
+      }
       else
+      {
         remAirSet.add(a);
+      }
     }
+    // We need to sort for the media file insertion step so that we can do a binary search.
+    Collections.sort(airset, DBObject.ID_COMPARATOR);
+    // We need to sort this one for later so we don't need to sort 1000's of smaller arrays.
+    Collections.sort(remAirSet, DBObject.ID_COMPARATOR);
+
     // We also need to be sure we analyze all of the files
     MediaFile[] mfs = wiz.getFiles();
-    for (int i = 0; i < mfs.length; i++)
+    for (int i = 0, mfsLen = mfs.length; i < mfsLen; i++)
     {
-      if (!mfs[i].archive && mfs[i].isTV())
-        airset.add(mfs[i].getContentAiring());
+      MediaFile mf = mfs[i];
+      if (mf != null && !mf.isArchiveFile() && mf.isTV())
+      {
+        Airing contentAir = mf.getContentAiring();
+        // Ensure we are inserting unique airings since we are not using a set.
+        int index = Collections.binarySearch(airset, contentAir, DBObject.ID_COMPARATOR);
+        if (index < 0)
+        {
+          index = -(index + 1);
+          airset.add(index, contentAir);
+        }
+      }
     }
 
-    Airing[] allAirs = airset.toArray(Pooler.EMPTY_AIRING_ARRAY);
-    Airing[] remAirs = remAirSet.toArray(Pooler.EMPTY_AIRING_ARRAY);
-
-    List<Agent> traitors = new ArrayList<Agent>();
-    allAgents = wiz.getAgents();
-    Set<Airing> watchedPotsToClear = new HashSet<Airing>();
-    if (Sage.DBG) System.out.println("CARNY Processing " + allAgents.length + " Agents & " + allAirs.length + " Airs");
-    boolean controlCPUUsage = doneInit && allAgents.length > 50 && Sage.getBoolean("control_profiler_cpu_usage", true);
-    // This array is used in the the Agent's calcWatchProb to avoid having to re-allocate a storage location each time we go through it.
-    // We make it bigger than what could possibly come up in the Agent's calc (i.e. every watch & waste matching it and none overlapping)
-    Airing[] airWorkCache = new Airing[1000 + wiz.getRawAccess(Wizard.WASTED_CODE, (byte) 0).length + wiz.getRawAccess(Wizard.WATCH_CODE, (byte) 0).length];
-    String paidProgRez = Sage.rez("Paid_Programming");
-    // We do a lot of String work for keyword favorites so have a buffer that they can do
-    // that work in to massively reduce GC'd memory allocations
-    StringBuffer sbCache = new StringBuffer();
-    String lastSplashMsg = null;
-    for (int i = 0; i < allAgents.length; i++)
+    // Add all of the work into a shared queue so we don't end up giving one thread a lot more work
+    // over another.
+    int submittedAgents;
+    if (!doneInit && Sage.getBoolean(LIMITED_CARNY_INIT, true))
```

Comment:
Ignore that comment. I look at the block again and this is deliberate. We only do limited init once. Why check for that condition for the rest of the runs of Carny? In my version we branch once instead of however many agents you have. I can't say I measured the performance here though, but I also didn't find it confusing. So I wasn't really swayed to make it simpler.